akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -0,0 +1,117 @@
+  }
+
+  def phasesFromConfig(conf: Config): Map[String, Set[String]] = {
+    import scala.collection.JavaConverters._
+    conf.root.unwrapped.asScala.toMap.map {
+      case (k, v: java.util.List[_]) ⇒ (k → v.asScala.map(_.toString).toSet)
+      case (k, v) ⇒
+        throw new IllegalArgumentException(s"Expected list value for [$k], got [$v]")
+    }
+  }
+
+  /**
+   * INTERNAL API: https://en.wikipedia.org/wiki/Topological_sorting
+   */
+  private[akka] def topologicalSort(phases: Map[String, Set[String]]): List[String] = {
+    var result = List.empty[String]
+    var unmarked = phases.keySet ++ phases.values.flatten // in case phase is not defined as key
+    var tempMark = Set.empty[String] // for detecting cycles
+
+    while (unmarked.nonEmpty) {
+      depthFirstSearch(unmarked.head)
+    }
+
+    def depthFirstSearch(u: String): Unit = {
+      if (tempMark(u))
+        throw new IllegalArgumentException("Cycle detected in graph of phases. It must be a DAG. " + phases)
+      if (unmarked(u)) {
+        tempMark += u
+        phases.getOrElse(u, Set.empty).foreach { v ⇒
+          depthFirstSearch(v)
+        }
+        unmarked -= u // permanent mark
+        tempMark -= u
+        result = u :: result
+      }
+    }
+
+    result.reverse
+  }
+
+}
+
+class CoordinatedShutdown(system: ExtendedActorSystem, phases: Map[String, Set[String]]) extends Extension {
+
+  private val knownPhases = phases.keySet ++ phases.values.flatten
+  private val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
+  private val tasks = new ConcurrentHashMap[String, Vector[() ⇒ Future[Done]]]
+  private val runStarted = new AtomicBoolean(false)
+  private val runPromise = Promise[Done]()
+
+  def addTask(phase: String)(task: () ⇒ Future[Done]): Unit = {
+    require(
+      knownPhases(phase),
+      s"unknown phase [$phase], known phases [$knownPhases]. All phases must be defined in configuration")
+    val current = tasks.get(phase)
+    if (current == null) {
+      if (tasks.putIfAbsent(phase, Vector(task)) != null)
+        addTask(phase)(task) // CAS failed, retry
+    } else {
+      if (!tasks.replace(phase, current, current :+ task))
+        addTask(phase)(task) // CAS failed, retry
+    }
+  }
+
+  def addNotification(phase: String, ref: ActorRef, message: Any): Unit = {
+    addTask(phase) {
+      () ⇒
+        ref ! message
+        Future.successful(Done)
+    }
+  }
+
+  def run(): Future[Done] = {
+    if (runStarted.compareAndSet(false, true)) {
+      import system.dispatcher
+      def loop(remainingPhases: List[String]): Future[Done] = {
+        remainingPhases match {
+          case Nil ⇒ Future.successful(Done)
+          case phase :: remaining ⇒
+            (tasks.get(phase) match {
+              case null ⇒ Future.successful(Done)
+              case tasks ⇒
+                // not that tasks within same phase are performed in parallel
+                Future.sequence(tasks.map(_.apply())).map(_ ⇒ Done)
```

Comment:
Yes, I have added timeout and recover. Recover (continue with next phase in case of failures) is enabled by default, but it is configurable per phase.